homeassistant/components/zwave_me/config_flow.py

```diff
@@ -0,0 +1,88 @@
+"""Config flow to configure ZWaveMe integration."""
+
+from dataclasses import asdict
+import logging
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+
+from .const import CONF_TOKEN, CONF_URL, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ZWaveMeConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """ZWaveMe integration config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def __init__(self):
+        """Initialize flow."""
+        self.url = vol.UNDEFINED
+        self.token = vol.UNDEFINED
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+
+        if self._async_current_entries():
+            return self.async_abort(reason="already_configured")
+
+        if user_input is not None:
+            if "host" in user_input:
+                self.url = user_input["host"]
+
+            else:
+                if "url" in user_input:
+                    self.url = user_input["url"]
+                else:
+                    user_input["url"] = self.url
+                self.token = user_input["token"]
+
+                if not user_input["url"].startswith("ws://") and not user_input[
+                    "url"
+                ].startswith("wss://"):
+                    user_input["url"] = "ws://" + user_input["url"] + ":8083"
+                    self.url = "ws://" + self.url + ":8083"
+
+                await self.async_set_unique_id(DOMAIN + self.url)
+                return self.async_create_entry(
```

Comment:
You didn't validate that the passed URL is correct.

Reply:
@lawfulchaos 

From @MartinHjelmare:
The best way would be to raise an exception in the library if it fails to connect. But since the library doesn't support asyncio and the connection happens in another thread that's not viable I think.

Maybe register a callback from the integration that is called by the on_open websocket callback. In the integration we can create an asyncio.Event and await this event with a timeout. The callback that we define in the integration, which is called on websocket open, can set the event. We can use hass.add_job in the callback to allow the library to set the event from another thread.

If the event isn't set within the timeout we raise ConfigEntryNotReady in the integration.

Reply:
You’re able to pass data back from sync to asyncio. All the connection initializer needs to know is if the connection has been established.

Inside asyncio you create a future that you pass to the sync part of the code. Here is some untested sample code. You create an asyncio future and set the result via a thread-safe loop function.

```
import asyncio

async def async_establish_connection(hass):
    established = asyncio.Future()
    loop = asyncio.get_running_loop()
    hass.add_job(
        connect,
        lambda result: loop.call_soon_threadsafe(established, result)
    )

    return await established

def connect(on_established):
    client = ...
    auth_success = ...

    if not auth_success:
        on_established(False)
        return

    on_established(True)
```